sophia/src/triple/stream/_error.rs

```diff
@@ -87,7 +87,54 @@ where
             SinkError(err) => err,
         }
     }
+    /// Switch source and sink error.
+    pub fn reverse(self) -> StreamError<SinkErr, SourceErr> {
+        match self {
+            SourceError(e) => SinkError(e),
+            SinkError(e) => SourceError(e),
+        }
+    }
 }
 
 /// Convenient type alias
 pub type StreamResult<T, E1, E2> = Result<T, StreamError<E1, E2>>;
+
+/// Extension trait for `SourceError`s.
+pub trait SourceResult<T, E1, E2>
+where
+    E1: 'static + Error,
+    E2: 'static + Error,
+{
+    /// Map an error to a `SourceError`.
+    fn source_err(self) -> StreamResult<T, E1, E2>;
```

Comment:
wouldn't `source_result` be a more appropriate name for this method?

Reply:
I thought about it but I decided to stay close to `map_err()` as this does nearly the same. If the `Result` yields an errors it maps/wraps it in an `SourceError`. In fact, this only replaces `.map_err(SourceError)`.

I think to write `fallible_function().source_err()?` makes it clear that if this function fails its error will be propagated as an `SourceError`. In difference, `fallible_function().map_err(SourceError)?` requires two steps of thinking:

1. Okay this will change the error-type
2. Ah so `SourceError` is a variant of `StreamError` and only wraps the given error.

Reply:
I totally agree with the added value of this trait, vs. having to explicitly call `map_err`. And the trait name is perfect for me.

Regarding the method name, I'm not entirely convinced, but I'm not entirely against it either... So I'm ok with keeping it.